rfcs/convergence/root-as-slot.md

```diff
@@ -0,0 +1,248 @@
+# RFC: root as a slot
+
+---
+
+_@bsunderhus @layershifter @ling1726_
+
+## Summary
+
+This RFC proposes to treat `root` as a _slot_ to simplify signature, improve `Typings` and avoid conflicts between `root` and other _slots_ signatures.
+
+Main changes would be:
+
+1. `root` becomes a regular slot on component state
+2. Support references on every _slot_
+3. Supports Custom Elements on every _slot_
+4. Allows a restricted `as` prop on every slot
+5. Stops using `getNativeElementProps` on every slot, will be used only for `root`
+
+## Problem statement
+
+There are disparities in how `root` is treated compared to _slots_. Some discussions have popped up due to these divergences between _slots_ and `root`:
+
+1. [Do not support as prop for slots that render native DOM elements] conflicts with [1st rule of ARIA opt-out mechanism]
+2. Support for Custom Elements on _slots_ but not for _root_
+3. Right now `Typings` for `root` aren't available as there's no way to split `root` from internal state.
+4. There's no possible way of passing refs to slots but `root` does
+5. [Primary slot]
+
+Major differences between `root` (shown below as 🌿) and _slots_ (shown below as 🎰):
+
+1. References 🌿
+2. Custom Elements as base element 🎰
+3. `as` prop 🌿
+4. Well defined Types 🎰
+
+## Detailed Design or Proposal
+
+For the user of the component perspective there will be no changes in the signature.
+
+### Converge `root` to a _slot_
+
+This RFC proposes to treat `root` as a _slot_.
+
+Treating `root` as a slot means having to declare `root` _Typings_ together with other _slots_. This will impact in a components state signature, where `root` will have to be declared.
+
+> ⚠️ That doesn't mean `root` will be available in component's properties interface
```

Comment:
Could you clarify this about root not being in the properties interface? It looks like `root` is in the `Slots` type in the example below. Is there some magic happening in the `ComponentProps<Slots>` helper to remove `root`?

Reply:
Or, if `root` is a prop, what happens if I write:
```jsx
<Component id="foo" root={{ id: "bar" }} />
```

Reply:
Yes, it's `ComponentProps` responsibility to remove `root` from `Slots` and flatten it in `Props` itself.

Reply:
> Or, if `root` is a prop, what happens if I write:
> 
> ```js
> <Component id="foo" root={{ id: "bar" }} />
> ```

You can't do that, since `root` is not available as a property